smdebug/profiler/profiler_config_parser.py

```diff
@@ -332,19 +341,83 @@ def write_tf_dataloader_flag(self, flag_filename):
 
         return success
 
+    def is_python_profiling_enabled(self):
+        return self.profiling_enabled and self.config.python_profiling_config.is_enabled()
+
+    def _handle_step_python_profiling(self, step_phase: StepPhase, mode: ModeKeys, current_step):
+        """Handle python profiling at the given step phase, mode and step by stopping python profiling and
+        starting python profiling again if python profiling is enabled and python profiling stats should be saved
+        for the current step.
+        """
+        if not self.is_python_profiling_enabled():
+            return
+
+        self.python_profiler.stop_profiling(
+            step_phase, end_mode=mode_keys_to_python_profile_mode(mode), end_step=current_step
+        )
+
+        if self.should_save_metrics(MetricsCategory.PYTHON_PROFILING, current_step):
+            self.python_profiler.start_profiling(
+                step_phase,
+                start_mode=mode_keys_to_python_profile_mode(mode),
+                start_step=current_step,
+            )
+
+    def start_pre_step_zero_python_profiling(self):
+        """Start pre-step zero python profiling if python profiling is enabled.
+        """
+        if not self.is_python_profiling_enabled():
+            return
+
+        return self.python_profiler.start_profiling(StepPhase.START)
+
+    def handle_step_start_python_profiling(self, mode: ModeKeys, current_step):
+        """Handle python profiling at the start of the step with the given mode and step by stopping python profiling
+        and starting python profiling again if python profiling stats should be saved for the current step.
+        """
+        return self._handle_step_python_profiling(StepPhase.STEP_START, mode, current_step)
+
+    def handle_step_end_python_profiling(self, mode: ModeKeys, current_step):
+        """Handle python profiling at the end of the step with the given mode and step by stopping python profiling
+        and starting python profiling again if python profiling stats should be saved for the current step.
+        """
+        return self._handle_step_python_profiling(StepPhase.STEP_END, mode, current_step)
+
+    def handle_forward_pass_end_python_profiling(self, mode: ModeKeys, current_step):
+        """Handle python profiling at the end of the forward pass with the given mode and step by stopping python
+        profiling and starting python profiling again if python profiling stats should be saved for the current step.
+        """
+        return self._handle_step_python_profiling(StepPhase.FORWARD_PASS_END, mode, current_step)
```

Comment:
Hooks shouldn't have any internal knowledge of python profiling and how it defines step phase, since it's framework agnostic. This should only fall under the scope of the profiler config parser, which will be the only component to have direct interaction with python profiling.

Reply:
At least, that's the purpose of the design change in this PR.